### ISSUE TITLE
Unable to move submodule resources to new submodule (not in state)

### ISSUE THREAD

**Issue by flo-sch** [hashicorp/terraform#24560]
Hello,

I am struggling a bit with submodules refactoring and general Terraform state management.
I am trying to move 3 **existing** resources of a submodule (`aws_s3_bucket` with related `aws_s3_bucket_public_access_block` and `aws_s3_bucket_policy`) to a **new** submodule (which is not yet in the state): https://github.com/terraform-aws-modules/terraform-aws-s3-bucket

Downtime is not affordable (production stage...), so destroying and re-creating is not an option. This is the main reason I am trying to do manual state management.

I have read the [`terraform state mv` documentation](https://www.terraform.io/docs/commands/state/mv.html), also found out some other issues that look similar (see below).

But even though `terraform state mv` seems to be successful, the new module is not referenced anywhere in the resulting state.
The initial resources are no longer referenced, but neither is the new module... So those resources seem  to be removed from the state and I am a bit lost in the process 😅

#### Scenario

To help understand a bit more, here are my resources addresses (full configuration files below)
(I am testing with one given {project_name} but I have tens of those to refactor)

Before refactoring:
- `module.project_1_frontend.module.s3.aws_s3_bucket.default`
- `module.project_1_frontend.module.s3.aws_s3_bucket_public_access_block.default`
- `module.project_1_frontend.module.s3.aws_s3_bucket_policy.default`

After refactoring (that new s3 module uses `count` internally, hence the `[0]`):
- `module.project_1_frontend.module.s3.module.s3_bucket.aws_s3_bucket.this[0]`
- `module.project_1_frontend.module.s3.module.s3_bucket.aws_s3_bucket_public_access_block.this[0]`
- `module.project_1_frontend.module.s3_policy.aws_s3_bucket_policy.default`

NOTE: the `aws_s3_bucket_policy` will use a different new module.

### Terraform Version

```
Terraform v0.12.24
+ provider.aws v2.53.0
```

#### Root project module

(Not touched during refactoring but just to help understanding)

```hcl
# {project_name}.tf
module "project_1_frontend" {
  source = "./modules/frontend"

  # [...]
}
```

#### Modules before refactoring

```hcl
######################
# My frontend module #
######################

# Contains other submodules like CloudFront, Route53 etc.
# modules/frontend/main.tf

module "s3" {
  source = "./modules/s3"

  # [...]
}


############################
# My frontend S3 submodule #
############################

# modules/frontend/modules/s3/main.tf
resource "aws_s3_bucket" "website" {
  # [...]
}

resource "aws_s3_bucket_public_access_block" "website_access_blocking" {
  # [...]
}

resource "aws_s3_bucket_policy" "s3_bucket_policy" {
  # [...]
}
```

#### Modules after refactoring

```hcl
############################
# My frontend module       #
# modules/frontend/main.tf #
############################

# Contains other submodules like CloudFront, Route53 etc.

module "s3" {
  source = "../shared/s3"

  # [...]
}

module "s3_policy" {
  source = "../shared/s3/policy"

  # [...]
}


#############################
# My shared S3 module       #
# modules/shared/s3/main.tf #
#############################

# modules/frontend/modules/s3/main.tf
module "s3_bucket" {
  source = "terraform-aws-modules/s3-bucket/aws"
  version = "~> 1.6"

  # [...]
}

# modules/shared/s3/policy/main.tf
resource "aws_s3_bucket_policy" "default" {
  # [...]
}
```

### Debug Output

(Sensible informations have been redacted)

https://gist.github.com/flo-sch/2a6f210d7292abac73051e562bf5d864

### Expected Behavior

1. The final state resulting of `terraform state mv` contains the new modules:
- `module.project_1_frontend.module.s3.module.s3_bucket` and its resources (`.aws_s3_bucket.this[0]` and `.aws_s3_bucket_public_access_block.this[0]`)
- `module.project_1_frontend.module.s3_policy.aws_s3_bucket_policy.default`

2. References to previous submodule resources are no longer referenced in the state (neither as resources nor as dependencies):
- `module.project_1_frontend.module.s3.module.aws_s3_bucket.default`
- `module.project_1_frontend.module.s3.module.aws_s3_bucket_public_access_block.default`
- `module.project_1_frontend.module.s3.aws_s3_bucket_policy.default`

3. Running `terraform plan` afterwards shows no changes to be applied

### Actual Behavior

1. The final state resulting of `terraform state mv` does not contain the new module or its resources

2. Previous submodule resources are no longer referenced as resources but their address is still present as dependencies for other resources (like my CloudFront submodule)

3. Running `terraform plan` afterwards shows the following changes:

> Terraform will perform the following actions:
>
>    module.project_1_frontend.module.cloudfront.aws_cloudfront_distribution.default will be updated in-place
>     _This is the distribution associated to the S3 bucket, which would get a new origin (the new S3 bucket which I hoped to have moved)_
> 	  **--> This would mean downtime in my case**
>
>   module.project_1_frontend.module.s3_policy.aws_s3_bucket_policy.default will be created
>   module.project_1_frontend.module.s3.module.s3_bucket.aws_s3_bucket.this[0] will be created
>   module.project_1_frontend.module.s3.module.s3_bucket.aws_s3_bucket_public_access_block.this[0] will be created
>
> Plan: 3 to add, 1 to change, 0 to destroy.

### Steps to Reproduce

I may be missing an obvious step here to somehow import the new module to the state, but I have no idea how to achieve that?

1. `terraform state pull > my-test-workspace.initial.tfstate`
2. `cp my-test-workspace.initial.tfstate my-test-workspace.final.tfstate`
3. `terraform state mv -state=my-test-workspace.final.tfstate -state-out=my-test-workspace.final.tfstate [...]` for the different resources
4. `terraform state push my-test-workspace.final.tfstate`
5. `terraform plan`

### Additional Context

I created a script chaining several `terraform state mv` (not sure if such practice is even legal? :D)

```sh
state="my-test-workspace"

# My plan is to add ~20 root modules to this array
modules=("project_1_frontend")

terraform state pull > $state.initial.tfstate

# Keep a copy of the initial plan to compare them
cp $state.initial.tfstate $state.final.tfstate

for module in ${modules[@]}; do
  echo "Migrating module: $module"

  terraform state mv -state=$state.final.tfstate -state-out=$state.final.tfstate "module.$module.module.s3.aws_s3_bucket.default" "module.$module.module.s3.module.s3_bucket.aws_s3_bucket.this[0]"
  terraform state mv -state=$state.final.tfstate -state-out=$state.final.tfstate "module.$module.module.s3.aws_s3_bucket_public_access_block.default" "module.$module.module.s3.module.s3_bucket.aws_s3_bucket_public_access_block.this[0]"
  terraform state mv -state=$state.final.tfstate -state-out=$state.final.tfstate "module.$module.module.s3.aws_s3_bucket_policy.default" "module.$module.module.s3_policy.aws_s3_bucket_policy.default"
done
```

### References

https://github.com/hashicorp/terraform/issues/21346
https://github.com/hashicorp/terraform/pull/23582
https://github.com/hashicorp/terraform/issues/23659
https://github.com/hashicorp/terraform/pull/22299

**Comment by mildwonkey** [comment]
I've tried to reproduce this issue and have been unsuccessful so far. It's possible that the problem has been fixed since this issue was opened, but it's also possible that my reproduction case was too simple to reproduce this issue. 

@flo-sch, sorry for the long silence here! If you are still experiencing issues with moving resources, can you provide a full  configuration that can been used as a reproduction case? There may be something about the dependencies between resources in your modules that I'm not thinking of which are leading to this behavior.

**Comment by flo-sch** [comment]
Hi @mildwonkey, do you need a repository or something?

(I have not re-tried this since the issue was opened)

**Comment by danieldreier** [comment]
@flo-sch can you please re-try this with the 0.13 RC1 release (or after Monday, the 0.13.0 GA release)? If you're still seeing this problem, the next thing we'd need is a full reproduction case that we can git clone and run, ideally something where all the steps are shell-scripted so it's totally unambiguous. I'm happy to iterate with you on exactly how that works - we just need something we can run without inventing or filling in any details.

**Comment by mildwonkey** [comment]
I am going to close this issue since I believe it's been fixed in v0.13. If you still see this problem in terraform v0.13 please open a new issue. Thanks!
